packages/core/test/app/ovm/deciders/and-decider.spec.ts

```diff
@@ -0,0 +1,283 @@
+        decider: cannotDecideDecider,
+        input: rightInput,
+      },
+      rightWitness,
+    }
+
+    try {
+      if (isDecide) {
+        await decider.decide(andInput, undefined)
+      } else {
+        await decider.checkDecision(andInput)
+      }
+      assert(false, 'This should throw a CannotDecideError')
+    } catch (e) {
+      if (!(e instanceof CannotDecideError)) {
+        assert(false, 'Error thrown should be CannotDecideError')
+      }
+    }
+  }
+
+  const testThrowWhenBothCannotDecide = async (isDecide: boolean = true) => {
+    const andInput: AndDeciderInput = {
+      left: {
+        decider: cannotDecideDecider,
+        input: leftInput,
+      },
+      leftWitness,
+      right: {
+        decider: cannotDecideDecider,
+        input: rightInput,
+      },
+      rightWitness,
+    }
+
+    try {
+      if (isDecide) {
+        await decider.decide(andInput, undefined)
+      } else {
+        await decider.checkDecision(andInput)
+      }
+      assert(false, 'This should throw a CannotDecideError')
+    } catch (e) {
+      if (!(e instanceof CannotDecideError)) {
+        assert(false, 'Error thrown should be CannotDecideError')
+      }
+    }
+  }
+
+  describe('decide', () => {
+    it('should return true with two true decisions', async () => {
+      await testReturnTrueWithTwoTrueDecisions()
+    })
+
+    it('should return false with left false', async () => {
+      await testReturnFalseWhenLeftIsFalse()
+    })
+
+    it('should return false with right false', async () => {
+      await testReturnFalseWhenRightIsFalse()
+    })
+
+    it('should return false with both false', async () => {
+      await testReturnFalseWhenBothAreFalse()
+    })
+
+    it('should throw when left cannot decide', async () => {
+      await testThrowWhenLeftCannotDecide()
+    })
+
+    it('should throw when right cannot decide', async () => {
+      await testThrowWhenRightCannotDecide()
+    })
+
+    it('should throw when both cannot decide', async () => {
+      await testThrowWhenBothCannotDecide()
+    })
+  })
+
+  describe('checkDecision', () => {
+    it('should return true with two true decisions', async () => {
+      await testReturnTrueWithTwoTrueDecisions(false)
+    })
+
+    it('should return false with left false', async () => {
+      await testReturnFalseWhenLeftIsFalse(false)
+    })
+
+    it('should return false with right false', async () => {
+      await testReturnFalseWhenRightIsFalse(false)
+    })
+
+    it('should return false with both false', async () => {
+      await testReturnFalseWhenBothAreFalse(false)
+    })
+
+    it('should throw when left cannot decide', async () => {
+      await testThrowWhenLeftCannotDecide(false)
+    })
+
+    it('should throw when right cannot decide', async () => {
+      await testThrowWhenRightCannotDecide(false)
+    })
+
+    it('should throw when both cannot decide', async () => {
+      await testThrowWhenBothCannotDecide(false)
+    })
```

Comment:
These test cases seem pretty comprehensive! Somewhat strange here how similar `decide(...)` and `checkDecision(...)` are, but still think it all makes sense at this point.

The only other tests I could think of were: 1) bad input values, and 2) nested AND predicates. That said, don’t think either are critical & personally would be happy with the current level of test coverage.